File: internal/tui/views/suppressions.go
package views

import (
	"context"
	"fmt"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/mailersend/mailersend-cli/internal/sdkclient"
	"github.com/mailersend/mailersend-cli/internal/tui/components"
	"github.com/mailersend/mailersend-cli/internal/tui/theme"
	"github.com/mailersend/mailersend-cli/internal/tui/types"
	"github.com/mailersend/mailersend-go"
)

// SuppressionType represents the type of suppression list.
type SuppressionType int

const (
	SuppressionBlocklist SuppressionType = iota
	SuppressionBounces
	SuppressionSpam
	SuppressionUnsubscribes
)

func (s SuppressionType) String() string {
	switch s {
	case SuppressionBlocklist:
		return "Blocklist"
	case SuppressionBounces:
		return "Hard Bounces"
	case SuppressionSpam:
		return "Spam Complaints"
	case SuppressionUnsubscribes:
		return "Unsubscribes"
	default:
		return "Unknown"
	}
}

var (
	tabStyle = lipgloss.NewStyle().
			Padding(0, 2)

	activeTabStyle = lipgloss.NewStyle().
			Bold(true).
			Foreground(theme.Primary).
			Background(theme.BgSelected).
			Padding(0, 2)

	tabBarStyle = lipgloss.NewStyle().
			Border(lipgloss.NormalBorder(), false, false, true, false).
			BorderForeground(theme.Muted).
			MarginBottom(1)
)

// SuppressionsView displays suppression lists.
type SuppressionsView struct {
	client        *mailersend.Mailersend
	transport     *sdkclient.CLITransport
	table         components.Table
	detail        components.DetailPanel
	items         []types.SuppressionItem
	loading       bool
	err           error
	width         int
	height        int
	focused       bool
	activeTab     SuppressionType
	tabFocused    bool
	showingDetail bool
}

// NewSuppressionsView creates a new suppressions view.
func NewSuppressionsView(client *mailersend.Mailersend, transport *sdkclient.CLITransport) SuppressionsView {
	columns := []components.Column{
		{Title: "EMAIL/PATTERN", Width: 35},
		{Title: "TYPE/REASON", Width: 25},
		{Title: "CREATED", Width: 19},
	}
	table := components.NewTable(columns)
	table.SetEmptyMessage("No suppression entries found.")

	return SuppressionsView{
		client:    client,
		transport: transport,
		table:     table,
		loading:   true,
		activeTab: SuppressionBlocklist,
	}
}

// SetClient updates the SDK client.
func (v *SuppressionsView) SetClient(client *mailersend.Mailersend, transport *sdkclient.CLITransport) {
	v.client = client
	v.transport = transport
}

// SetSize sets the view dimensions.
func (v *SuppressionsView) SetSize(width, height int) {
	v.width = width
	v.height = height
	// Reserve space for tab bar
	v.table.SetSize(width, height-4)
}

// SetFocused sets whether this view is focused.
func (v *SuppressionsView) SetFocused(focused bool) {
	v.focused = focused
	v.table.SetFocused(focused && !v.tabFocused)
}

// Loading returns whether the view is loading.
func (v SuppressionsView) Loading() bool {
	return v.loading
}

// Error returns any error.
func (v SuppressionsView) Error() error {
	return v.err
}

// ItemCount returns the number of items.
func (v SuppressionsView) ItemCount() int {
	return len(v.items)
}

// Fetch returns a command to fetch suppressions based on active tab.
func (v SuppressionsView) Fetch() tea.Cmd {
	return func() tea.Msg {
		if v.client == nil {
			return types.SuppressionsLoadedMsg{Err: nil}
		}

		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		var items []types.SuppressionItem
		var err error

		switch v.activeTab {
		case SuppressionBlocklist:
			items, err = v.fetchBlocklist(ctx)
		case SuppressionBounces:
			items, err = v.fetchBounces(ctx)
		case SuppressionSpam:
			items, err = v.fetchSpam(ctx)
		case SuppressionUnsubscribes:
			items, err = v.fetchUnsubscribes(ctx)
		}

		return types.SuppressionsLoadedMsg{
			Items: items,
			Type:  v.activeTab.String(),
			Err:   err,
		}
	}
}

func (v *SuppressionsView) fetchBlocklist(ctx context.Context) ([]types.SuppressionItem, error) {
	result, _, err := v.client.Suppression.ListBlockList(ctx, nil)
	if err != nil {
		return nil, sdkclient.WrapError(v.transport, err)
	}

	var items []types.SuppressionItem
	for _, b := range result.Data {
		items = append(items, types.SuppressionItem{
			ID:        b.ID,
			Pattern:   b.Pattern,
			Type:      b.Type,
			CreatedAt: b.CreatedAt.Format(time.RFC3339),
		})
	}
	return items, nil
}

func (v *SuppressionsView) fetchBounces(ctx context.Context) ([]types.SuppressionItem, error) {
	result, _, err := v.client.Suppression.ListHardBounces(ctx, nil)
	if err != nil {
		return nil, sdkclient.WrapError(v.transport, err)
	}

	var items []types.SuppressionItem
	for _, b := range result.Data {
		items = append(items, types.SuppressionItem{
			ID:        b.ID,
			Pattern:   b.Recipient.Email,
			Reason:    b.Reason,
			CreatedAt: b.CreatedAt.Format(time.RFC3339),
		})
	}
	return items, nil
}

func (v *SuppressionsView) fetchSpam(ctx context.Context) ([]types.SuppressionItem, error) {
	result, _, err := v.client.Suppression.ListSpamComplaints(ctx, nil)
	if err != nil {
		return nil, sdkclient.WrapError(v.transport, err)
	}

	var items []types.SuppressionItem
	for _, s := range result.Data {
		items = append(items, types.SuppressionItem{
			ID:        s.ID,
			Pattern:   s.Recipient.Email,
			Reason:    "Spam complaint",
			CreatedAt: s.CreatedAt.Format(time.RFC3339),
		})
	}
	return items, nil
}

func (v *SuppressionsView) fetchUnsubscribes(ctx context.Context) ([]types.SuppressionItem, error) {
	result, _, err := v.client.Suppression.ListUnsubscribes(ctx, nil)
	if err != nil {
		return nil, sdkclient.WrapError(v.transport, err)
	}

	var items []types.SuppressionItem
	for _, u := range result.Data {
		items = append(items, types.SuppressionItem{
			ID:        u.ID,
			Pattern:   u.Recipient.Email,
			Reason:    u.ReadableReason,
			CreatedAt: u.CreatedAt.Format(time.RFC3339),
		})
	}
	return items, nil
}

// Update handles messages for this view.
func (v SuppressionsView) Update(msg tea.Msg) (SuppressionsView, tea.Cmd) {
	switch msg := msg.(type) {
	case types.SuppressionsLoadedMsg:
		v.loading = false
		v.err = msg.Err
		if msg.Err == nil {
			v.items = msg.Items
			v.updateTable()
		}
	}
	return v, nil
}

// HandleKey handles key events when this view is active.
func (v *SuppressionsView) HandleKey(msg tea.KeyMsg) tea.Cmd {
	// Handle detail view navigation
	if v.showingDetail {
		switch msg.String() {
		case "esc", "backspace", "q":
			v.showingDetail = false
		}
		return nil
	}

	switch msg.String() {
	case "j", "down":
		v.table.MoveDown()
	case "k", "up":
		v.table.MoveUp()
	case "g":
		v.table.GotoTop()
	case "G":
		v.table.GotoBottom()
	case "h", "left":
		v.prevTab()
		v.loading = true
		return v.Fetch()
	case "l", "right":
		v.nextTab()
		v.loading = true
		return v.Fetch()
	case "enter":
		v.showDetail()
	case "r":
		v.loading = true
		v.table.SetLoading(true)
		return v.Fetch()
	}
	return nil
}

// SelectedItem returns the currently selected suppression item.
func (v SuppressionsView) SelectedItem() *types.SuppressionItem {
	idx := v.table.Cursor()
	if idx >= 0 && idx < len(v.items) {
		return &v.items[idx]
	}
	return nil
}

func (v *SuppressionsView) showDetail() {
	item := v.SelectedItem()
	if item == nil {
		return
	}

	v.detail.SetTitle("Suppression Entry")

	created := item.CreatedAt
	if t, err := time.Parse(time.RFC3339, item.CreatedAt); err == nil {
		created = t.Format("2006-01-02 15:04:05")
	}

	rows := []components.DetailRow{
		{Label: "ID", Value: item.ID},
		{Label: "Email/Pattern", Value: item.Pattern},
		{Label: "List Type", Value: v.activeTab.String()},
	}

	if item.Type != "" {
		rows = append(rows, components.DetailRow{Label: "Type", Value: item.Type})
	}
	if item.Reason != "" {
		rows = append(rows, components.DetailRow{Label: "Reason", Value: item.Reason})
	}

	rows = append(rows, components.DetailRow{Label: "Created", Value: created})

	v.detail.SetRows(rows)
	v.detail.SetSize(v.width, v.height-4)
	v.showingDetail = true
}

func (v *SuppressionsView) nextTab() {
	v.activeTab = (v.activeTab + 1) % 4
}

func (v *SuppressionsView) prevTab() {
	v.activeTab--
	if v.activeTab < 0 {
		v.activeTab = 3
	}
}

func (v *SuppressionsView) updateTable() {
	var rows [][]string
	for _, item := range v.items {
		created := item.CreatedAt
		if t, err := time.Parse(time.RFC3339, item.CreatedAt); err == nil {
			created = t.Format("2006-01-02 15:04:05")
		}

		typeOrReason := item.Type
		if item.Reason != "" {
			typeOrReason = item.Reason
		}

		rows = append(rows, []string{
			item.Pattern,
			typeOrReason,
			created,
		})
	}
	v.table.SetRows(rows)
	v.table.SetLoading(false)
}

// View renders the suppressions view.
func (v SuppressionsView) View() string {
	if v.showingDetail {
		return v.detail.View()
	}

	var b strings.Builder

	// Tab bar
	tabs := []SuppressionType{
		SuppressionBlocklist,
		SuppressionBounces,
		SuppressionSpam,
		SuppressionUnsubscribes,
	}

	var tabViews []string
	for _, tab := range tabs {
		style := tabStyle
		if tab == v.activeTab {
			style = activeTabStyle
		}
		tabViews = append(tabViews, style.Render(tab.String()))
	}

	tabBar := lipgloss.JoinHorizontal(lipgloss.Top, tabViews...)
	b.WriteString(tabBarStyle.Render(tabBar))
	b.WriteString("\n")

	// Hint for tab navigation
	hint := fmt.Sprintf("← → to switch tabs | %d items", len(v.items))
	b.WriteString(lipgloss.NewStyle().Foreground(theme.Muted).Render(hint))
	b.WriteString("\n\n")

	// Table
	b.WriteString(v.table.View())

	return b.String()
}

// ShowingDetail returns whether the detail view is active.
func (v SuppressionsView) ShowingDetail() bool {
	return v.showingDetail
}

File: src/tui/views/suppressions.rs
use anyhow::Result;
use crossterm::event::KeyCode;
use ratatui::layout::{Constraint, Layout, Rect};
use ratatui::style::{Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::Paragraph;
use ratatui::Frame;
use serde_json::Value;

use super::{fmt_datetime, KeyResult};
use crate::api::Client;
use crate::tui::components::detail::{DetailPanel, DetailRow};
use crate::tui::components::table::{Cell, Column, Table};
use crate::tui::theme;
use crate::util::{jpath, jstr};

pub const EMPTY: &str = "No suppression entries found.";

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Tab {
    Blocklist,
    HardBounces,
    SpamComplaints,
    Unsubscribes,
}

impl Tab {
    pub const ALL: [Tab; 4] = [
        Tab::Blocklist,
        Tab::HardBounces,
        Tab::SpamComplaints,
        Tab::Unsubscribes,
    ];

    pub fn label(self) -> &'static str {
        match self {
            Tab::Blocklist => "Blocklist",
            Tab::HardBounces => "Hard Bounces",
            Tab::SpamComplaints => "Spam Complaints",
            Tab::Unsubscribes => "Unsubscribes",
        }
    }

    fn path(self) -> &'static str {
        match self {
            Tab::Blocklist => "/suppressions/blocklist",
            Tab::HardBounces => "/suppressions/hard-bounces",
            Tab::SpamComplaints => "/suppressions/spam-complaints",
            Tab::Unsubscribes => "/suppressions/unsubscribes",
        }
    }

    pub fn next(self) -> Tab {
        Tab::ALL[(self as usize + 1) % Tab::ALL.len()]
    }

    pub fn prev(self) -> Tab {
        Tab::ALL[(self as usize + Tab::ALL.len() - 1) % Tab::ALL.len()]
    }
}

pub struct Item {
    pub id: String,
    pub pattern: String,
    pub type_: String,
    pub reason: String,
    pub created_at: String,
}

pub fn columns() -> Vec<Column> {
    vec![
        Column {
            title: "EMAIL/PATTERN",
            width: 35,
        },
        Column {
            title: "TYPE/REASON",
            width: 25,
        },
        Column {
            title: "CREATED",
            width: 19,
        },
    ]
}

pub fn fetch(client: &Client, tab: Tab) -> Result<Vec<Item>> {
    let body = client.get(tab.path())?;
    let data: Vec<Value> = body
        .get("data")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();

    Ok(data
        .iter()
        .map(|d| {
            let (pattern, type_, reason) = match tab {
                Tab::Blocklist => (jstr(d, "pattern"), jstr(d, "type"), String::new()),
                Tab::HardBounces => (
                    jpath(d, "recipient.email"),
                    String::new(),
                    jstr(d, "reason"),
                ),
                Tab::SpamComplaints => (
                    jpath(d, "recipient.email"),
                    String::new(),
                    "Spam complaint".to_string(),
                ),
                Tab::Unsubscribes => (
                    jpath(d, "recipient.email"),
                    String::new(),
                    jstr(d, "readable_reason"),
                ),
            };
            Item {
                id: jstr(d, "id"),
                pattern,
                type_,
                reason,
                created_at: fmt_datetime(&jstr(d, "created_at")),
            }
        })
        .collect())
}

pub struct View {
    pub table: Table,
    detail: DetailPanel,
    pub items: Vec<Item>,
    pub loading: bool,
    pub active_tab: Tab,
    showing_detail: bool,
}

impl View {
    pub fn new() -> Self {
        let mut table = Table::new(columns(), EMPTY);
        table.set_loading(true);
        Self {
            table,
            detail: DetailPanel::new(),
            items: Vec::new(),
            loading: true,
            active_tab: Tab::Blocklist,
            showing_detail: false,
        }
    }

    pub fn set_loaded(&mut self, items: Vec<Item>) {
        let rows = items
            .iter()
            .map(|item| {
                let type_or_reason = if item.reason.is_empty() {
                    item.type_.clone()
                } else {
                    item.reason.clone()
                };
                vec![
                    Cell::plain(item.pattern.clone()),
                    Cell::plain(type_or_reason),
                    Cell::plain(item.created_at.clone()),
                ]
            })
            .collect();
        self.items = items;
        self.table.set_rows(rows);
        self.table.set_loading(false);
        self.loading = false;
    }

    pub fn handle_key(&mut self, code: KeyCode) -> KeyResult {
        if self.showing_detail {
            if matches!(code, KeyCode::Esc | KeyCode::Backspace | KeyCode::Char('q')) {
                self.showing_detail = false;
            }
            return KeyResult::None;
        }

        match code {
            KeyCode::Char('j') | KeyCode::Down => self.table.move_down(),
            KeyCode::Char('k') | KeyCode::Up => self.table.move_up(),
            KeyCode::Char('g') => self.table.goto_top(),
            KeyCode::Char('G') => self.table.goto_bottom(),
            KeyCode::Char('h') | KeyCode::Left => {
                self.active_tab = self.active_tab.prev();
                self.loading = true;
                return KeyResult::Fetch;
            }
            KeyCode::Char('l') | KeyCode::Right => {
                self.active_tab = self.active_tab.next();
                self.loading = true;
                return KeyResult::Fetch;
            }
            KeyCode::Enter => self.show_detail(),
            KeyCode::Char('r') => {
                self.loading = true;
                self.table.set_loading(true);
                return KeyResult::Fetch;
            }
            _ => {}
        }
        KeyResult::None
    }

    fn show_detail(&mut self) {
        let Some(item) = self.items.get(self.table.cursor()) else {
            return;
        };
        self.detail.set_title("Suppression Entry");
        let mut rows = vec![
            DetailRow::new("ID", item.id.clone()),
            DetailRow::new("Email/Pattern", item.pattern.clone()),
            DetailRow::new("List Type", self.active_tab.label()),
        ];
        if !item.type_.is_empty() {
            rows.push(DetailRow::new("Type", item.type_.clone()));
        }
        if !item.reason.is_empty() {
            rows.push(DetailRow::new("Reason", item.reason.clone()));
        }
        rows.push(DetailRow::new("Created", item.created_at.clone()));
        self.detail.set_rows(rows);
        self.showing_detail = true;
    }

    pub fn render(&mut self, frame: &mut Frame, area: Rect, focused: bool) {
        if self.showing_detail {
            frame.render_widget(Paragraph::new(self.detail.lines()), area);
            return;
        }

        let [chrome, table_area] =
            Layout::vertical([Constraint::Length(4), Constraint::Min(0)]).areas(area);
        frame.render_widget(Paragraph::new(self.tab_chrome()), chrome);

        self.table.set_focused(focused);
        self.table
            .set_size(table_area.width as usize, table_area.height as usize);
        frame.render_widget(Paragraph::new(self.table.lines()), table_area);
    }

    fn tab_chrome(&self) -> Vec<Line<'static>> {
        let mut tabs: Vec<Span<'static>> = Vec::new();
        let mut bar_width = 0usize;
        for tab in Tab::ALL {
            let label = format!("  {}  ", tab.label());
            bar_width += label.chars().count();
            let style = if tab == self.active_tab {
                Style::new()
                    .fg(theme::PRIMARY)
                    .bg(theme::BG_SELECTED)
                    .add_modifier(Modifier::BOLD)
            } else {
                Style::new()
            };
            tabs.push(Span::styled(label, style));
        }

        vec![
            Line::from(tabs),
            Line::styled("─".repeat(bar_width), Style::new().fg(theme::MUTED)),
            Line::styled(
                format!("← → to switch tabs | {} items", self.items.len()),
                Style::new().fg(theme::MUTED),
            ),
            Line::raw(""),
        ]
    }
}
